File: extras/Python/cmd_messenger/queue/command_queue.py
"""``CommandQueue`` — abstract base. Direct port of C# ``CommandQueue``.

Each queue owns an :class:`AsyncWorker`. Concrete subclasses implement
:meth:`_process_queue` which is the worker job.
"""
from __future__ import annotations

import threading
from abc import ABC, abstractmethod
from typing import List

from ..async_worker import AsyncWorker
from .command_strategy import CommandStrategy
from .general_strategy import GeneralStrategy
from .list_queue import ListQueue


class CommandQueue(ABC):
    """Abstract base for the send/receive command queues."""

    def __init__(self) -> None:
        self._queue: ListQueue[CommandStrategy] = ListQueue()
        self._queue_lock = threading.Lock()
        self._general_strategies: List[GeneralStrategy] = []
        self._worker = AsyncWorker(self._process_queue, name=self.__class__.__name__)

    # ------------------------------------------------------------------
    # Worker passthrough (mirrors C# properties)
    # ------------------------------------------------------------------
    @property
    def is_running(self) -> bool:
        return self._worker.is_running

    @property
    def is_suspended(self) -> bool:
        return self._worker.is_suspended

    @property
    def count(self) -> int:
        """Number of pending commands. NOT thread-safe (matches C# note)."""
        return len(self._queue)

    @property
    def is_empty(self) -> bool:
        return len(self._queue) == 0

    def clear(self) -> None:
        with self._queue_lock:
            self._queue.clear()

    def add_general_strategy(self, general_strategy: GeneralStrategy) -> None:
        general_strategy.command_queue = self._queue
        self._general_strategies.append(general_strategy)

    @abstractmethod
    def queue_command(self, command_strategy: CommandStrategy) -> None:
        """Queue the wrapped strategy. Subclass-specific behaviour."""

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------
    def start(self) -> None:
        self._worker.start()

    def stop(self) -> None:
        self._worker.stop()
        self.clear()

    def suspend(self) -> None:
        self._worker.suspend()

    def resume(self) -> None:
        self._worker.resume()

    def _signal_worker(self) -> None:
        self._worker.signal()

    def dispose(self) -> None:
        if self._worker.state.name != "STOPPED":
            self.stop()

    def __enter__(self) -> "CommandQueue":
        self.start()
        return self

    def __exit__(self, *exc_info) -> None:
        self.dispose()

    # ------------------------------------------------------------------
    # Worker job — subclass-supplied
    # ------------------------------------------------------------------
    @abstractmethod
    def _process_queue(self) -> bool:
        """Return ``True`` if more work pending, ``False`` to wait for signal."""

File: extras/TypeScript/src/queue/commandQueue.ts
import { Command } from '../command.js';
import { CommandStrategy } from './commandStrategy.js';
import { GeneralStrategy } from './generalStrategy.js';
import { ListQueue } from './listQueue.js';

export abstract class CommandQueue<TCommand extends Command = Command> {
  protected readonly queue = new ListQueue<CommandStrategy<TCommand>>();
  protected readonly generalStrategies: GeneralStrategy[] = [];
  private running = false;
  private suspended = false;
  private processing = false;
  private signalPending = false;

  get isRunning(): boolean {
    return this.running;
  }

  get isSuspended(): boolean {
    return this.suspended;
  }

  get count(): number {
    return this.queue.count;
  }

  get isEmpty(): boolean {
    return this.queue.isEmpty;
  }

  clear(): void {
    this.queue.clear();
  }

  addGeneralStrategy(generalStrategy: GeneralStrategy): void {
    generalStrategy.commandQueue = this.queue as ListQueue<CommandStrategy>;
    this.generalStrategies.push(generalStrategy);
  }

  abstract queueCommand(commandStrategy: CommandStrategy<TCommand>): void;

  start(): void {
    this.running = true;
    this.signalWorker();
  }

  stop(): void {
    this.running = false;
    this.clear();
  }

  suspend(): void {
    this.suspended = true;
  }

  resume(): void {
    this.suspended = false;
    this.signalWorker();
  }

  dispose(): void {
    this.stop();
  }

  protected signalWorker(): void {
    if (!this.running || this.suspended || this.processing || this.signalPending) {
      return;
    }
    this.signalPending = true;
    queueMicrotask(() => {
      this.signalPending = false;
      void this.drain();
    });
  }

  protected abstract processQueue(): Promise<boolean> | boolean;

  private async drain(): Promise<void> {
    if (this.processing || !this.running || this.suspended) {
      return;
    }
    this.processing = true;
    try {
      while (this.running && !this.suspended) {
        const hasMore = await this.processQueue();
        if (!hasMore) {
          break;
        }
      }
    } finally {
      this.processing = false;
    }
    if (this.running && !this.suspended && !this.isEmpty) {
      this.signalWorker();
    }
  }
}
